Throttle requests if the translation backlog hits the limit

mmaslankaprv opened the pull request:
Introduced a component that is responsible for throttling Iceberg
enabled producers when partition translation backlog target is not met.

The `datalake_throttle_manager` keeps track of all Iceberg related
producers. The producer is marked as iceberg enabled on any shard, then
the manager map reduces the state and throttling is always calculated on
shard 0.

In order to skip the x-shard call for non-iceberg producers and in case
the throttling is not required, every shard local manager can verify if
the throttling is required. Only if the translation backlog are over the
limit the x-shard call is required to verify if the producer is iceberg
enabled one. (shard 0 is the only one that has producer state from all
other shards).

The throttling is calculated based on the time since the translation
error was reported. Currently the max throttle will be applied after 5
minutes if translation problem persists.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
k/bazel: extracted kafka loggers to separate bazel library

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
dl_manager: added methods allowing querying translation status

The method return number of partitions which backlog target is not met
and the number of partitions for which translation is blocked.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
kafka: introduced datalake throttle manager

Introduced a component that is responsible for throttling Iceberg
enabled producers when partition translation backlog target is not met.

The `datalake_throttle_manager` keeps track of all Iceberg related
producers. The producer is marked as iceberg enabled on any shard, then
the manager map reduces the state and throttling is always calculated on
shard 0.

In order to skip the x-shard call for non-iceberg producers and in case
the throttling is not required, every shard local manager can verify if
the throttling is required. Only if the translation backlog are over the
limit the x-shard call is required to verify if the producer is iceberg
enabled one. (shard 0 is the only one that has producer state from all
other shards).

The throttling is calculated based on the time since the translation
error was reported. Currently the max throttle will be applied after 5
minutes if translation problem persists.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
k/server: wire in datalake throttle manager

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
k/produce: mark iceberg related producer

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
k/connection_ctx: added iceberg contribution to produce delay

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/dt

mmaslankaprv commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--44b993a6-d091-4fe1-b6d3-99d3f10d082b-->
<!--TEST RESULTS-->
<!--f2b9abe9-0d3c-49d8-a584-854021827cda-->
<!--TEST RESULTS-->
<!--47e4ad74-7a33-4d30-975c-3d4a406bfb94-->
<!--TEST RESULTS-->
<!--230c8df8-644f-4258-8ae5-a263f4c62f79-->
<!--TEST RESULTS-->
<!--666f6216-6a00-44ab-b523-4d90823ac30b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61960">build#61960</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| translator_test_rpfixture.translator_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61960#0195197e-1149-4975-bcaa-5cdd3016c0b0 | FAIL | 0/2 |
| translator_test_rpfixture.translator_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61960#0195197e-114a-47ef-a2ff-6ea903f22d32 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62003">build#62003</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62003#01951d9a-2ef0-437d-876b-9ffcd55412f0 | FLAKY | 1/2 |
| rptest.tests.crash_loop_checks_test.CrashLoopChecksTest.test_crash_report_with_signal.signo=Signals.SIGABRT.signal_shard=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62003#01951d7f-8677-4c8b-9667-b53183ee58da | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62003#01951d9a-2ef0-437d-876b-9ffcd55412f0 | FLAKY | 1/3 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62003#01951d9a-2ef0-437d-876b-9ffcd55412f0 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63120">build#63120</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63120#019593ff-18a1-4e3d-855c-9c7183962f37 | FLAKY | 1/12 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/63120#019593ff-18a2-46f7-bd8d-f758a93b0083 | FLAKY | 1/6 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/63120#019593ff-18a3-4808-a978-daeb56c8bedb | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63120#019593ff-18a4-4be0-b2b2-b52fd64f2994 | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/63120#019593ff-18a1-4e3d-855c-9c7183962f37 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/63120#019593ff-18a2-46f7-bd8d-f758a93b0083 | FLAKY | 1/15 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/63120#01959402-da0e-499a-a986-0752da6ed7aa | FLAKY | 1/3 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_spec_evolution.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/63120#019593ff-18a1-4e3d-855c-9c7183962f37 | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/63120#01959402-da0f-4637-b02b-2ea206379889 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63220">build#63220</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63220#0195a3bb-e138-4128-8de6-d032bb43b57a | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63220#0195a3cf-2c1e-426d-8fe6-9b121eda8dcc | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/63220#0195a3bb-e139-4a30-9b9c-b7666e873ca1 | FLAKY | 1/11 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63234">build#63234</a></summary>


| test_id                                                                                                                                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63234#0195a4f0-5262-45a3-93ab-d641113059f8 | FLAKY         | 1/2      |
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63234#0195a50a-b355-4995-9184-944261907218 | FLAKY         | 1/2      |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_spec_evolution_rpcn.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63234#0195a50a-b354-491f-b6cb-7b37c0779187 | FLAKY         | 1/2      |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63234#0195a50a-b355-4995-9184-944261907218 | FLAKY         | 1/2      |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63234#0195a50a-b355-4995-9184-944261907218 | FLAKY         | 1/2      |
</details>
<!--/body-->

mmaslankaprv committed:
k/quota_manager: made gc_interval runtime configurable

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/dt

mmaslankaprv committed:
datalake: made max concurrent translations runtime configurable

By making the max concurrent translations runtime configurable we make
it easier to test the scheduling or in extreme cases test how Redpanda
would behave if translation is completely stopped (by setting the max
concurrent translations property to `0`)

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
dl/tests: added basic test validating if datalake throttling works

Added a test that validates if the Iceberg related producers are
correctly throttled if translation can not keep up.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

WillemKauf commented on src/v/kafka/server/datalake_throttle_manager.cc:
```diff
@@ -0,0 +1,231 @@
+     * Only run update timer on shard 0
+     */
+    if (ss::this_shard_id() == 0) {
+        _update_timer.arm_periodic(100ms);
```
nit: should this timer be on a tunable interval?

mmaslankaprv replied:
Actually i think it is one of the properties that we would set and forget about

WillemKauf commented on src/v/kafka/server/datalake_throttle_manager.cc:
```diff
@@ -0,0 +1,231 @@
+        merge_producer_maps(_global_producers, shard_map);
+    }
+
+    std::erase_if(_global_producers, [this](const auto& p) {
```
Why are we erasing from `_global_producers` here? Since this map seems to want to track all producers, it feels like it would be a bit easier to parse this logic if it were in `get_producer_throttle()`

mmaslankaprv replied:
I wanted to make the `get_producer_throttle` as efficient as possible as it is called in the hot path of processing produce requests, this is why i put gc in the background. Let me think about how would that look like in `get_producer_throttle`

WillemKauf commented on src/v/kafka/server/datalake_throttle_manager.cc:
```diff
@@ -0,0 +1,231 @@
+    auto time_in_degraded_state
+      = std::chrono::duration_cast<std::chrono::milliseconds>(
+        ss::lowres_clock::now() - _last_no_issues_timestamp);
+    auto max_throttle_ratio = double(time_in_degraded_state.count()) / 300000;
```
nit: can move the definition of these variables into the `if()`, unused otherwise

WillemKauf commented on src/v/config/configuration.cc:
```diff
@@ -4011,9 +4011,9 @@ configuration::configuration()
+      {.needs_restart = needs_restart::no, .visibility = visibility::tunable},
       4,
-      {.min = 1, .max = 8})
+      {.min = 0, .max = 8})
```
If reducing this minimum to 0 is primarily for use in testing, check out [__REDPANDA_TEST_DISABLE_BOUNDED_PROPERTY_CHECKS](https://github.com/redpanda-data/redpanda/pull/24544) instead of opening this up to users.

mmaslankaprv replied:
TIL

mmaslankaprv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/63120-->
<!--41c0ae9f-f08d-496b-8f44-9cd777a61be9-->
<!--title-->
## Retry command for [Build#63120](https://buildkite.com/redpanda/redpanda/builds/63120)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/datalake_upgrade_test.py::DatalakeUpgradeTest.test_upload_through_upgrade@{"cloud_storage_type":1,"query_engine":"spark"}
```
<!--/body-->

ztlpn commented on src/v/redpanda/application.h:
```diff
@@ -195,6 +195,7 @@ class application {
     ss::sharded<rpc::connection_cache> _connection_cache;
     ss::sharded<kafka::group_manager> _group_manager;
     ss::sharded<experimental::cloud_topics::app> _reconciler;
+    ss::sharded<kafka::datalake_throttle_manager> datalake_throttle_manager;
```
nit: missing leading `_`

mmaslankaprv replied:
It is public variable hence the name without `_`

ztlpn commented on src/v/kafka/server/app.cc:
```diff
@@ -42,6 +42,7 @@ seastar::future<> server_app::init(
   seastar::sharded<cluster::security_frontend>& sec,
   seastar::sharded<cluster::controller_api>& ctrl,
   seastar::sharded<cluster::tx_gateway_frontend>& tx,
+  seastar::sharded<datalake_throttle_manager>& itm,
```
nit: itm -> dtm?

ztlpn commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -436,4 +436,21 @@ ss::future<uint64_t> datalake_manager::disk_usage() {
+ */
+size_t datalake_manager::partitions_with_translation_blocked() const {
+    // TODO: Return blocked if partition wasn't translated for a long time f.e.
+    // moret
```
Was the comment cropped? Is always returning 0 here okay?

mmaslankaprv replied:
I think it is ok for now. The idea is to update this piece when we figure out the heuristics for blocked translation partitions/

ztlpn commented on src/v/kafka/server/server.h:
```diff
@@ -231,6 +237,8 @@ class server final
     // processing incoming requests.
     ss::scheduling_group get_request_handler_sg() const;
 
+    void mark_datalake_producer(const std::optional<std::string_view>&);
```
nit: param name for reader's convenience

ztlpn commented on src/v/kafka/server/server.cc:
```diff
@@ -457,6 +460,14 @@ ss::future<> server::apply(ss::lw_shared_ptr<net::connection> conn) {
 
+void server::mark_datalake_producer(
+  const std::optional<std::string_view>& client_id) {
+    if (!config::shard_local_cfg().iceberg_enabled()) {
```
Should we just directly check that `_datalake_throttle_manager` is initialized? We've had a lot of crashes in the past related to services being unexpectedly not initialized.

ztlpn commented on src/v/kafka/server/server.h:
```diff
@@ -138,6 +139,11 @@ class server final
     quota_manager& quota_mgr() { return _quota_mgr.local(); }
+
+    datalake_throttle_manager& datalake_throttle_manager() {
+        return _datalake_throttle_manager.local();
```
Same concern about DTM being unexpectedly uninitialized - I see that this method is used only in 1 place, maybe encapsulate it in `server::maybe_throttle_producer()` and don't expose the DTM directly?

ztlpn commented on src/v/kafka/server/connection_context.h:
```diff
@@ -286,6 +286,9 @@ class connection_context final
       get_scheduling_group_override(api_key) const;
 
+    ss::future<std::chrono::milliseconds>
+    get_iceberg_producer_throttle(const std::optional<std::string_view>&);
```
nit: param name

ztlpn commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -567,14 +567,24 @@ produce_topic(produce_ctx& octx, produce_request::topic& topic) {
+              // if topic is iceberg enabled update iceberg throttle manager.
+              if (is_iceberg_enabled) {
+                  octx.rctx.server().local().mark_datalake_producer(
+                    octx.rctx.header().client_id);
```
I wonder if client_id is the right granularity to track producers? AFAICS it is just some string like `rdkafka` which seems a bit coarse.

ztlpn commented on tests/rptest/tests/datalake/throttling_test.py:
```diff
@@ -0,0 +1,119 @@
+
+            # Set the target lag to min value of 10s
+            rpk.alter_topic_config(self.topic_name,
+                                   "redpanda.iceberg.target.lag.ms", "10000")
```
nit: this seems to be done by DatalakeServices these days?

ztlpn commented on src/v/kafka/server/datalake_throttle_manager.h:
```diff
@@ -0,0 +1,111 @@
+
+    status_provider_fn _shard_status_provider;
+    config::binding<std::chrono::milliseconds> _producer_gc_threshold;
+    config::binding<std::chrono::milliseconds> _max_kafka_throttle;
```
nit: would be nice to not mix members and member functions in the order they appear in the class definition

ztlpn commented on src/v/kafka/server/datalake_throttle_manager.cc:
```diff
@@ -0,0 +1,232 @@
+} // namespace
+void datalake_throttle_manager::merge_producer_maps(
+  producers_map_t& target, const producers_map_t& source) {
+    for (const auto& [client_id, state] : source) {
```
should we insert yields when merging producing maps or are they expected to be small?

mmaslankaprv replied:
can we do it in subsequent PR ?

ztlpn replied:
sure, sg

ztlpn commented on src/v/kafka/server/datalake_throttle_manager.cc:
```diff
@@ -0,0 +1,232 @@
+
+std::chrono::milliseconds
+datalake_throttle_manager::calculate_throttle() const {
+    if (_translation_status.partitions_translation_blocked > 0) {
```
nit: shard assert here as well to assist the reader in understanding which methods are called only on shard 0? Or maybe just merge with the previous method

ztlpn commented on src/v/kafka/server/datalake_throttle_manager.cc:
```diff
@@ -0,0 +1,232 @@
+        auto time_in_degraded_state
+          = std::chrono::duration_cast<std::chrono::milliseconds>(
+            ss::lowres_clock::now() - _last_no_issues_timestamp);
+        auto max_throttle_ratio = double(time_in_degraded_state.count())
```
should we clamp `max_throttle_ratio` by 1.0 here?

ztlpn reviewed (commented):
Looks good, mostly nits, but I had a concern about using `client_id` for marking datalake producers.
